### PR TITLE
Rapl node capping daemon

### PR DESCRIPTION
**Pull request by ridv**
* Initial work on RAPL daemon.
* Python code has been ported over.
* Tests for daemon have been added.

Left to do in separate PRs:

* Systemd unit file needs to be added.
*  Scheduler code when calling capper should be rewired to send an HTTP payload instead of starting an expensive SSH connection.

### CODE REVIEW

**Comment by ridv** [comment]
Maybe we can put this in a separate repo since there's no dependencies on the main project.

File: rapl-daemon/README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# RAPL Daemon
+
+This runs a server that is capable changing the percentage at which
```
Comment (pradykaushik) [inline rapl-daemon/README.md]: capable **of** changing

File: rapl-daemon/main.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	"encoding/json"
+	"fmt"
+	"html"
+	"log"
+	"net/http"
+)
+
+const powercapDir = "/sys/class/powercap/"
+
+// Cap is a payload that is expected from Elektron to cap a node
```
Comment (pradykaushik) [inline rapl-daemon/main.go]: missing period at the end of the sentence.

File: rapl-daemon/main.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+type Cap struct {
+	Percentage int
+}
+
+func main() {
+	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, "Unsupported endpoint %s", html.EscapeString(r.URL.Path))
```
Comment (pradykaushik) [inline rapl-daemon/main.go:27]: error messages should be lowercase.

File: rapl-daemon/main.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+	})
+
+	http.HandleFunc("/powercap", powercapEndpoint)
+	log.Fatal(http.ListenAndServe(":9090", nil))
+}
+
+// Handler powercapping HTTP API endpoint
```
Comment (pradykaushik) [inline rapl-daemon/main.go]: missing period at end of sentence.

File: rapl-daemon/util.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	"math"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+const raplPrefixCPU = "intel-rapl"
+
+// constraint_0 is usually the longer window while constraint_1 is usually the longer window
```
Comment (pradykaushik) [inline rapl-daemon/util.go]: constraint_1 is the shorter window.

File: rapl-daemon/util.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+const maxPowerFileShortWindow = "constraint_0_max_power_uw"
+const powerLimitFileShortWindow = "constraint_0_power_limit_uw"
+
+// capNode uses pseudo files made available by the Linux kernel
+// in order to capNode CPU power. More information is available at:
+// https://www.kernel.org/doc/html/latest/power/powercap/powercap.html
+func capNode(base string, percentage int) error {
+
+	if percentage <= 0 || percentage > 100 {
+		return fmt.Errorf("cap percentage must be between 0 (non-inclusive) and 100 (inclusive): %d", percentage)
```
Comment (pradykaushik) [inline rapl-daemon/util.go]: maybe just say **cap percentage must be in the range (0, 100]**.

File: rapl-daemon/util.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+const raplPrefixCPU = "intel-rapl"
+
+// constraint_0 is usually the longer window while constraint_1 is usually the longer window
+const maxPowerFileShortWindow = "constraint_0_max_power_uw"
```
Comment (pradykaushik) [inline rapl-daemon/util.go]: constraint_0 is the longer window.
Going by [this](https://github.com/spdfg/elektron/blob/master/scripts/RAPL_PKG_Throttle.py#L48) we are intending to use the longer window for running average calculations.
So, change the variable name to `maxPowerFileLongWindow`?

File: rapl-daemon/util.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+const powerLimitFileShortWindow = "constraint_0_power_limit_uw"
```
Comment (pradykaushik) [inline rapl-daemon/util.go]: ditto.

File: rapl-daemon/util.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+		}
+
+		if fields[0] == raplPrefixCPU {
+			maxPower, err := maxPower(filepath.Join(base, file.Name(), maxPowerFileShortWindow))
+			if err != nil {
+				fmt.Println("unable to retreive max power for zone ", err)
+				continue
+			}
+
+			// We use floats to mitigate the possibility of an integer overflows.
```
Comment (pradykaushik) [inline rapl-daemon/util.go]: overflow~s~

File: rapl-daemon/util.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+				continue
+			}
+		}
+	}
+
+	return nil
+}
+
+// maxPower returns the value in float of the maximum watts a power zone
+// can use.
```
Comment (pradykaushik) [inline rapl-daemon/util.go]: fits on a single line.

File: rapl-daemon/util.go [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+		// Fields should be in the form intel-rapl:X where X is the power zone
+		// We ignore sub-zones which follow the form intel-rapl:X:Y
+		if len(fields) != 2 {
+			continue
+		}
+
+		if fields[0] == raplPrefixCPU {
+			maxPower, err := maxPower(filepath.Join(base, file.Name(), maxPowerFileShortWindow))
+			if err != nil {
+				fmt.Println("unable to retreive max power for zone ", err)
```
Comment (pradykaushik) [inline rapl-daemon/util.go:47]: any reason for just printing the error and not returning it?
Basically, the question is whether we need to continue to attempt to powercap if an error occurs anywhere in the process.

File: rapl-daemon/util.go [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+				continue
+			}
+
+			// We use floats to mitigate the possibility of an integer overflows.
+			powercap := uint64(math.Ceil(float64(maxPower) * (float64(percentage) / 100)))
+
+			err = capZone(filepath.Join(base, file.Name(), powerLimitFileShortWindow), powercap)
+			if err != nil {
+				fmt.Println("unable to write powercap value: ", err)
```
Comment (pradykaushik) [inline rapl-daemon/util.go:56]: same here regarding whether the error to continue if error occurs or should we just return it.

File: rapl-daemon/util_test.go [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+	os.Exit(m.Run())
+}
+
+// TODO(rdelvalle): Create filesystem only once and allow tests to use it
+func TestCapNode(t *testing.T) {
+	err := capNode(raplDir, 95)
+	assert.NoError(t, err)
+
+	t.Run("badPercentage", func(t *testing.T) {
```
Comment (pradykaushik) [inline rapl-daemon/util_test.go]: I think it's safe to use spaces ("bad percentage") as the words are hyphenated by default.
Although, I do not mind the camelcasing.

File: rapl-daemon/util_test.go [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+	assert.Equal(t, maxWattage, maxWatts)
+
+	t.Run("nameDoesNotExist", func(t *testing.T) {
+		_, err := maxPower("madeupname")
+		assert.Error(t, err)
+	})
+}
+
+func TestCapZone(t *testing.T) {
+	const percentage float64 = .50
```
Comment (pradykaushik) [inline rapl-daemon/util_test.go:86]: The payload declares `percentage` as an integer. So, we would never hit this scenario.
Should the percentage in the payload be an integer?

**Review by pradykaushik — CHANGES_REQUESTED** [review]
I agree that we can place the rapl-daemon code in a separate repo.
I have commented on a few things. Apart from that, everything looks okay to me.

File: rapl-daemon/main.go (hunk 3 above)
Comment (ridv) [reply rapl-daemon/main.go:27]: In this case this is some http content being served back to to the client that made the request. I did make it lower case though.

File: rapl-daemon/util.go (hunk 9 above)
Comment (ridv) [reply rapl-daemon/util.go]: Can you expand on this comment?

File: rapl-daemon/util.go (hunk 11 above)
Comment (ridv) [reply rapl-daemon/util.go:47]: Agreed, I don't know what the right path to take is here. If the file write fails, we could be left in an inconsistent state. The only options here are to bail and return an error right away, or to continue trying to cap the node and then returning an array of zones where were unable to cap the node in.

File: rapl-daemon/util_test.go (hunk 14 above)
Comment (ridv) [reply rapl-daemon/util_test.go:86]: The payload does indeed take an integer, but when we calculate powercap, I'd rather multiply by a fraction than to multiply by a (0,100] integer and then divide by 100. This is done in the next line and it's just for testing.

File: rapl-daemon/util.go (hunk 11 above)
Comment (pradykaushik) [reply rapl-daemon/util.go:47]: Any chance we can roll back the power limits set to the previous values if and when any of the file writes fail?

File: rapl-daemon/util.go (hunk 6 above)
Comment (pradykaushik) [reply rapl-daemon/util.go]: Actually, your way of representing the ranges is more explicit and clear. One is not well-versed with range expressions would still be able to understand your explanation. So, ignore my previous statement and leave it as is.

File: rapl-daemon/util.go (hunk 9 above)
Comment (pradykaushik) [reply rapl-daemon/util.go]: Either change **overflows** to **overflow**, or get rid of the "an".

File: rapl-daemon/util.go (hunk 11 above)
Comment (ridv) [reply rapl-daemon/util.go:47]: If the write fails, we're sort of left in limbo. If the write to modify it failed, then most likely the write to restore the value will fail too. We can try reading it and seeing if that's the same value as before.

File: rapl-daemon/util.go (hunk 5 above)
Comment (ridv) [reply rapl-daemon/util.go]: Good catch, fixed.

File: rapl-daemon/util.go (hunk 7 above)
Comment (ridv) [reply rapl-daemon/util.go]: Fixed this, definitely a mistake on my part. Thanks for catching it!

File: rapl-daemon/util.go (hunk 6 above)
Comment (ridv) [reply rapl-daemon/util.go]: I'll revert it

File: rapl-daemon/util_test.go (hunk 13 above)
Comment (ridv) [reply rapl-daemon/util_test.go]: Good point, fixed.

**Review by pradykaushik — APPROVED** [review]

**Pull request merged**